components/browser/session-storage/src/test/java/mozilla/components/browser/session/storage/serialize/BrowserStateWriterReaderTest.kt

```diff
@@ -147,6 +152,135 @@ class BrowserStateWriterReaderTest {
         assertEquals(tab.content.url, restoredTab.historyMetadata!!.url)
     }
 
+    @Test
+    fun `Read and write tab ensuring its restored flag is set correction`() {
+        val engineState = createFakeEngineState()
+        val engine = createFakeEngine(engineState)
+
+        val tab = createTab(
+            url = "https://www.mozilla.org",
+            title = "Mozilla",
+            contextId = "work",
+            source = SessionState.Source.External.CustomTab(
+                caller = ExternalPackage("com.mozilla.test", PackageCategory.PRODUCTIVITY)
+            )
+        )
+        assertFalse(tab.restored)
+
+        val writer = BrowserStateWriter()
+        val reader = BrowserStateReader()
+
+        val file = AtomicFile(
+            File.createTempFile(UUID.randomUUID().toString(), UUID.randomUUID().toString())
+        )
+
+        assertTrue(writer.writeTab(tab, file))
+
+        val restoredTab = reader.readTab(engine, file)
+        assertNotNull(restoredTab!!)
+
+        assertTrue(restoredTab.restored)
+    }
+
+    @Test
+    fun `Read and write tab with external custom tab source and full caller`() {
```

Comment:
Once this is in Fenix it would be nice to add a new test case to `RestoringBrowsingSessionsTest`, which tests actual snapshots from actual installs. :)